### DISCUSSION TITLE
Package/@Languages in WiX 4

### DISCUSSION THREAD

**Discussion by avivanoff** [wixtoolset/issues#7325] [Questions]
In WiX 3 there is Package/@Languages element to specify a list of language the MSI package supports. When using wix convert tool to upgrade to WiX4, this element just gets removed. How do you specify languages in WiX 4 project?

Also posted on [StackOverflow](https://stackoverflow.com/questions/75879120/package-languages-in-wix-4).

**Comment by BMurri** [comment]
Short answer: Package/@Language. A single package can only contain one language. However, if you are using any wixtoolset provided assets (such as the WixUI extension), there are better ways. Those better ways are better even if you are not using WixUI.

If you are using [`MSBuild`](https://wixtoolset.org/docs/tools/msbuild/) (aka `dotnet`, `Visual Studio`), best practice is to use `.wxl` files, where you set your language, and if you include several, several packages will be created. Second best practice is to provide a list of cultures you are building (MSBuild property `Cultures`).

If not using MSBuild, provide the culture using the `-culture` parameter and wxl files using the `-loc` parameter to the [`wix`](https://wixtoolset.org/docs/tools/wixexe/) tool. Repeat -loc as needed.

BTW, specifying the cultures to build from the command-line/through MSBuild was considered best-practice during v3, as well.

**Comment by avivanoff** [comment]
Not true. A single MSI package may contain multiple embedded transforms for different languages. Hence the [summary information](https://learn.microsoft.com/en-us/windows/win32/msi/template-summary) can contain multiple language id’s. We used to set it via Package/@Languages in WIX 3. I need a way to do the same in WIX 4.

**Reply by robmen** [reply]
From the documentation you sent:

> Merge Modules are the only packages that may have multiple languages. Only one language can be specified in a source installer database. For more information, see [Multiple Language Merge Modules](https://learn.microsoft.com/en-us/windows/win32/msi/multiple-language-merge-modules).

Is setting multiple languages in WiX v3 actually doing anything for your MSI? It may just be invalid data slipping through that is now caught in WiX v4.

**Reply by avivanoff** [reply]
We are using [Multi-Language MSI Packages](http://www.installsite.org/pages/en/msi/articles/embeddedlang/).

**Reply by BMurri** [reply]
I know all about it. Yes, I've used it before. No, it's never been supported by Microsoft, and it's never been directly supported by the WIX Toolset (as far as I recall).

**Reply by avivanoff** [reply]
I can do it in WIX 3 using Package/@Languages.

All supported/unsupported/undocumented arguments aside, shouldn’t [SummaryInformation element](https://wixtoolset.org/docs/schema/wxs/summaryinformation/) have support for setting languages?

**Reply by robmen** [reply]
Not in v4 because setting multiple languages isn't supported by the Windows Installer. It was a bug that v3 allowed it.

That said, in future versions of WiX, we could decide to implement this undocumented feature and add the necessary tests to validate it works. The Windows Installer is not changing at this time, so we _may_ decide to lean more heavily on observed behavior vs. documented behavior. For example, MSI cabinet filenames are supposed to be 8.3, but it seems that limitation is outdated. So perhaps in a future release we will change the default `MediaTemplate` value to the "name of the output MSI" + diskid + ".cab". But not in v4.

**Reply by BMurri** [reply]
Also, the command-line to the `wix` tool sets that value in summary information for you, as does do values in `.wxl` files, in the same way the platform values are set in summary information from the tool's command-line. If you use MSBuild, it calls the WIX tool for you.

So, yes, it's settable from the Toolset. Just not with multiple languages, except for merge modules, at this time, as Rob mentioned.

Like I responded elsewhere, as well as from the guide you shared, there are tools from Microsoft that you have to use anyway to embed your transforms that will let you set those values.

**Reply by BradyJ9** [reply]
> Not in v4 because setting multiple languages isn't supported by the Windows Installer. It was a bug that v3 allowed it.

I'm not sure if this is necessarily the case.  https://learn.microsoft.com/en-us/windows/win32/msi/manage-substorages  This Microsoft documentation itself details a script that is for embedding transforms, which a common use case is for localizing the installer (see tutorial at the bottom).  I too am missing this feature in Wix v4 and am realizing there's no other way to set the Summary Information property for multiple supported languages.

This did work for us in v3 so I'm wondering why it was removed too, as it seemed a crucial part to localizing a single installer.

**Reply by BradyJ9** [reply]
I stand corrected.  That tutorial seems to outline a different way of localizing that is "best practice" that involves modifying the tables with localized versions.  That being said the old way seems far easier, and there seem to be sufficient Microsoft tools in place to allow this way.  It just makes for a rocky transition in Wix v4 to have this feature pulled out from us.

EDIT: But here again, as I type this, I find more official Windows documentation that seems to suggest that transforming your installer to be localized is supported and common practice: https://learn.microsoft.com/en-us/windows/win32/msi/merges-and-transforms

I can't seem to understand what the best practice is, but does this not seem like Wix v4 removed a common use case?

**Reply by barnson** [reply]
To build a package with embedded language transforms, you have to use another tool and can modify the summary information stream at that time. Until/unless WiX supports such packages natively, it makes no sense to support multiple languages in a package.

**Reply by BradyJ9** [reply]
Except for the fact that it was supported in v3, and now v4 has dropped a crucial feature. Many installers that relied on this for localization are now broken in those languages without a hint of documentation explaining why or the workaround. That seems like an issue.

**Reply by jmairboeck** [reply]
It was "supported" in Wix v3 in a way that resulted in broken MSI packages without some post-processing (creating and embedding language transforms into the MSI). Otherwise the MSIs coming out of Wix would not be runnable.

Now, with v4, the post-processing needs to do an additional step, that is, adding the language ID to the summary information [Template property](https://learn.microsoft.com/en-us/windows/win32/msi/template-summary), as shown in the link posted by @avivanoff. Afterwards, it works just as before. Additionally, the "intermediate" output MSIs created by Wix are also runnable as is.

I have created a small tool that integrates the functionality of the Windows Installer SDK scripts into a single executable that can be run as a post-build step. Maybe I should publish that somewhere ...

**Reply by BMurri** [reply]
To sum this all up: the supported, recommended way to do localized installers is to build each language installer, pick one to be the base, generate transforms from the base for all the others, then ship the base and all relevant transforms, and include the selected transform with the base MSI in the installation command-line. This almost asks for a bootstrapper to make sure the command line isn't messed up by the end customer. Then again, using a bootstrapper may mean that the need to localize the MSI completely goes away (which is my personal recommendation for everyone who can get away with it, which I'm pretty sure is almost everyone except certain people shipping exclusively B2B software with specific deployment needs involving MSI advertising, which I haven't seen in a contract for almost 2 decades, now).

The "trick" of embedding the transforms with very specific names and turning the summary information language setting into a language list was never documented nor supported by Microsoft, which is why it's not currently supported by WiX, and is the reason tests for it were never added to the Toolset, and is the reason that the v3 behavior is considered to be in error. It's true that this "trick" is "easier" than the recommended supported way, that it was discovered by someone reverse engineering installation packages released by a different department at Microsoft, and it's unlikely (though not impossible) that Microsoft will break that use case, all reasons many people choose to prepare their packages that way. However, if something does go wrong, Microsoft is not likely to promise to fix it either, without you spending a bunch of money on them.

Since added steps/tools have always been required above and beyond what has been included in the WiX Toolset to take advantage of that "trick", it shouldn't be a huge hurdle to migrate your build tooling to WiX v4 and still end up with the desired final result. You will simply have one additional step in your process to add.

Once again, there's no bug here in v4 (although it's obvious that there was one in v3). The invitation to request support for this specific use case in some future version of the Toolset was extended, so if that will be a benefit to you, please feel free to create an issue to request support for that use case end-to-end so that all the steps will be done by the Toolset for you. Since this is open source software, if you want to ensure that it'll be in some future version of the Toolset, feel free to include a pull request with tests, or hire someone to provide that PR, to go along with that issue (I know Firegiant sells support contracts, and I think they still offer connecting you to independent contractors). There's an entire community here that would help you with your PR if it's that important for you.

**Reply by jmairboeck** [reply]
I published our tool here, if anyone needs it: https://github.com/RISCSoftware/CreateEmbedLangTransform

**Comment by BMurri** [comment]
@avivanoff From the link you provided:

> Merge Modules are the only packages that may have multiple languages. Only one language can be specified in a source installer database. For more information, see [Multiple Language Merge Modules](https://learn.microsoft.com/en-us/windows/win32/msi/multiple-language-merge-modules).

However, since you are already doing things that are not supported by Windows Installer (even though they wrote the code that makes it work), go ahead and use the [sdk tools](https://learn.microsoft.com/en-us/windows/win32/msi/windows-installer-development-tools) to do the job (aka `msiinfo`)
